servant/src/Servant/API/UVerb.hs

```diff
@@ -0,0 +1,123 @@
+import Servant.API.Status (KnownStatus, statusVal)
+import Servant.API.UVerb.OpenUnion
+
+class KnownStatus (StatusOf a) => HasStatus (a :: *) where
+  type StatusOf (a :: *) :: Nat
+
+statusOf :: forall a proxy. HasStatus a => proxy a -> Status
+statusOf = const (statusVal (Proxy :: Proxy (StatusOf a)))
+
+instance KnownStatus n => HasStatus (WithStatus n a) where
+  type StatusOf (WithStatus n a) = n
+
+-- | If an API can respond with 'NoContent' we assume that this will happen
+-- with the status code 204 No Content. If this needs to be overridden,
+-- 'WithStatus' can be used.
+instance HasStatus NoContent where
+  type StatusOf NoContent = 204
+
+class HasStatuses (as :: [*]) where
+  type Statuses (as :: [*]) :: [Nat]
+  statuses :: Proxy as -> [Status]
+
+instance HasStatuses '[] where
+  type Statuses '[] = '[]
+  statuses _ = []
+
+instance (HasStatus a, HasStatuses as) => HasStatuses (a ': as) where
+  type Statuses (a ': as) = StatusOf a ': Statuses as
+  statuses _ = statusOf (Proxy :: Proxy a) : statuses (Proxy :: Proxy as)
+
+newtype WithStatus (k :: Nat) a = WithStatus a
+  deriving (Eq, Show, GHC.Generic)
+
+instance (GHC.Generic (WithStatus n a), ToJSON a) => ToJSON (WithStatus n a)
+
+instance (GHC.Generic (WithStatus n a), FromJSON a) => FromJSON (WithStatus n a)
+
+instance MimeRender ctype a => MimeRender ctype (WithStatus _status a) where
+  mimeRender contentTypeProxy (WithStatus a) = mimeRender contentTypeProxy a
+
+instance MimeUnrender ctype a => MimeUnrender ctype (WithStatus _status a) where
+  mimeUnrender contentTypeProxy input =
+    WithStatus <$> mimeUnrender contentTypeProxy input
+
+-- FUTUREWORK:
+-- @type Verb method statusCode contentTypes a = UVerb method contentTypes [WithStatus statusCode a]@
+-- no, wait, this is not the same.  this would mean people would have to use 'respond' instead
+-- of 'pure' or 'return'.
+data UVerb (method :: StdMethod) (contentTypes :: [*]) (as :: [*])
+
+type Union = NS Identity
+
+-- | Convenience function to apply a function to an unknown union element using a type class.
+-- All elements of the union must have instances in the type class, and the function is
+-- applied unconditionally.
+--
+-- See also: 'extractUResp'.
+collapseUResp ::
+  forall (c :: * -> Constraint) (a :: *) (as :: [*]).
+  All c as =>
+  Proxy c -> (forall x. c x => x -> a) -> Union as -> a
+collapseUResp proxy render = collapse_NS . cmap_NS proxy (K . render . runIdentity)
+
+-- | Convenience function to extract a union element using 'cast', ie. return the value if the
+-- selected type happens to be the actual type of the union in this value, or 'Nothing'
+-- otherwise.
+--
+-- See also: 'collapseUResp'.
+extractUResp ::
+  forall (a :: *) (as :: [*]).
+  (All Typeable as, Typeable a, IsMember a as) =>
+  Union as -> Maybe a
+extractUResp = collapse_NS . cmap_NS (Proxy @Typeable) (K . cast . runIdentity)
```

Comment:
WE shouldn't use typeable here? it sounds very hacky. We have all the type level info already.

Reply:
This is just a _very_ complicated way of 

``
match  :: NS f as -> Maybe (f a)
```

which is part of the `IsMember` typeclass ;)

Reply:
This is still relevant

Reply:
I decided I want the helper to remain here, because it does have a nicer type.  But you're right, the implementation had potential of improvement.  :)  cb008f2b09c042f5ccd9331f47eeef1be1e4b88